file.go

```diff
@@ -714,6 +718,22 @@ func getBufioReader(r io.Reader) *bufio.Reader {
 func putBufioReader(rbuf *bufio.Reader) {
 	if rbuf != nil {
 		rbuf.Reset(nil)
-		bufioReaderPool.Put(rbuf)
+		getBufioReaderPool(rbuf.Size()).Put(rbuf)
+	}
+}
+
+func getBufioReaderPool(size int) *sync.Pool {
+	bufioReaderPoolLock.RLock()
+	if pool := bufioReaderPool[size]; pool != nil {
+		bufioReaderPoolLock.RUnlock()
+		return pool
 	}
+	bufioReaderPoolLock.RUnlock()
```

Comment:
How about simplifying the synchronization logic to use a simple mutex? Access to this global variable should be infrequent and the critical section is short, plus `sync.RWMutex` still contends on an atomic value internally so it scales poorly with multiple cores (which I assume would be the intent of using it here).

Reply:
No issues swapping to `sync.Mutex`, it does simplify the logic, at a likely low cost.

Do the number of cores matter? Couldn't a goroutine yield in the critical section and block other attempts to `getBufioReaderPool()` even on a single core?

Also, unsure if this matters, but we are considering a design which would likely have thousands of parquet files in a single process. We already have a hundreds.

Reply:
> Do the number of cores matter? Couldn't a goroutine yield in the critical section and block other attempts to getBufioReaderPool() even on a single core?

Goroutines might be preempted even when running on a single CPU (or with GOMAXPROCS=1), so there may be contention on the mutex no matter how many cores there are, tho it would be more pronounced with large number of cores.

I do think that access to this variable will be so infrequent in comparison with the rest of the code executed by a program that it would be really unlikely that we would ever measure contention on this global mutex. If it ever becomes a problem we can figure it out then, and we'll have data to verify that the change actually improves parallelism.